### PR TITLE
[11.0][IMP] contract: adapted to purchase contracts

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Based on https://github.com/OCA/contract/pull/97.

### CODE REVIEW

**Comment by mreficent** [comment]
@jbeficent can you review?

**Comment by mreficent** [comment]
@jbeficent can you review?

**Comment by MiquelRForgeFlow** [comment]
@jbeficent , can yo review now?

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Tested functionally, including migration script

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza maybe you want to review this?

File: contract/views/account_analytic_account_view.xml [hunk 1]
```diff
@@ -39,7 +42,7 @@
                             />
                 </div>
                 <group col="4" attrs="{'invisible': [('recurring_invoices','!=',True)]}">
-                    <field name="contract_template_id" colspan="4"/>
+                    <field name="contract_template_id" colspan="4" domain="['|', ('contract_type', '=', contract_type), ('contract_type', '=', False)]"/>
```
Comment (RoelAdriaans) [inline contract/views/account_analytic_account_view.xml]: When you create a new contract template from the purchase contract, the type is a sale contract. Please provide an default_contract_type=contract_type context

File: contract/tests/test_contract.py [hunk 2]
```diff
@@ -269,11 +275,11 @@ def test_contract_onchange_product_id_name(self):
...
 
     def test_contract_count(self):
-        """It should return contract count."""
-        count = self.partner.contract_count + 2
+        """It should return sale contract count."""
+        count = self.partner.sale_contract_count + 2
         self.contract.copy()
         self.contract.copy()
-        self.assertEqual(self.partner.contract_count, count)
+        self.assertEqual(self.partner.sale_contract_count, count)
```
Comment (RoelAdriaans) [inline contract/tests/test_contract.py:291]: Please add an test for purchase_contract_count , ideally on the same partner to have different counts

**Review by RoelAdriaans — CHANGES_REQUESTED** [review]
For the contract template, can you add the contract type to the tree view, and maybe group-by filter?

https://github.com/Eficent/contract/blob/bad0c8309450f6fd211640f5fba000ddf13735b2/contract/views/account_analytic_contract_view.xml#L65

**Comment by MiquelRForgeFlow** [comment]
@RoelAdriaans comments attended

**Review by RoelAdriaans — APPROVED** [review]

**Comment by RoelAdriaans** [comment]
Code and functional :+1: 
Build is red because of missing tests..
Lot of the changes are from res.partner where there are new lines for just creating a view. Creating a test for that looks pointless? This file is now at 62% coverage.. 

Other item is the `_onchange_contract_type` in `account_analytic_contract.py` , adding a test for that could be helpful.

File: contract/__manifest__.py [hunk 3]
```diff
@@ -8,15 +8,15 @@
...
+    'version': '11.0.4.0.0',
     'category': 'Contract Management',
     'license': 'AGPL-3',
     'author': "OpenERP SA, "
               "Tecnativa, "
               "LasLabs, "
               "Odoo Community Association (OCA)",
     'website': 'https://github.com/oca/contract',
-    'depends': ['base', 'account', 'analytic'],
+    'depends': ['base', 'account_invoicing', 'analytic'],
```
Comment (pedrobaeza) [inline contract/__manifest__.py]: Why this change?

File: contract/migrations/11.0.4.0.0/post-migration.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Eficent - Miquel Raïch
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+def migrate(cr, version):
+    """All previous contract where of 'sale' type."""
```
Comment (pedrobaeza) [inline contract/migrations/11.0.4.0.0/post-migration.py]: s/where/were

File: contract/migrations/11.0.4.0.0/post-migration.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    if not version:
+        return
+    cr.execute(
+        """
+        UPDATE account_analytic_account
+        SET contract_type = 'sale'
```
Comment (pedrobaeza) [inline contract/migrations/11.0.4.0.0/post-migration.py]: I'm thinking you don't need this putting as default value `sale` in the field `contract_type`

File: contract/models/account_analytic_contract.py [hunk 6]
```diff
@@ -26,6 +25,12 @@ class AccountAnalyticContract(models.Model):
         comodel_name="res.partner",
         string="Partner (always False)",
     )
+    contract_type = fields.Selection(
+        selection=[
+            ('sale', _('Sale')),
```
Comment (pedrobaeza) [inline contract/models/account_analytic_contract.py]: Selection fields don't require to translate the label this way (they are extracted automatically), so remove `_` calls

File: contract/models/account_analytic_contract.py [hunk 7]
```diff
@@ -73,11 +79,22 @@ class AccountAnalyticContract(models.Model):
         default=lambda self: self.env.user.company_id,
     )
 
+    @api.onchange('contract_type')
+    def _onchange_contract_type(self):
+        if self.contract_type == 'purchase':
+            for line in self.recurring_invoice_line_ids.filtered(
```
Comment (pedrobaeza) [inline contract/models/account_analytic_contract.py]: `self.recurring_invoice_line_ids.filtered('automatic_price').update({'automatic_price': False})`

File: contract/models/res_partner.py [hunk 8]
```diff
@@ -7,30 +7,63 @@
...
         compute='_compute_contract_count',
     )
 
     def _compute_contract_count(self):
-        Contract = self.env['account.analytic.account']
+        contract_model = self.env['account.analytic.account']
         today = fields.Date.today()
         for partner in self:
-            partner.contract_count = Contract.search_count([
+            partner.sale_contract_count = contract_model.search_count([
```
Comment (pedrobaeza) [inline contract/models/res_partner.py]: This is not very performance wise. Please implement it through a `read_group` outside the loop with lazy=False, the domain without the contract_type, and grouping by partner_id and contract_type, and you'll get the same results but faster.

File: contract/models/res_partner.py [hunk 9]
```diff
@@ -7,30 +7,63 @@
...
+            partner.purchase_contract_count = contract_model.search_count([
+                ('recurring_invoices', '=', True),
+                ('partner_id', 'child_of', partner.ids),
+                '|',
+                ('date_end', '=', False),
+                ('date_end', '>=', today),
+                ('contract_type', '=', 'purchase'),
+            ])
+
+    def act_show_sale_contract(self):
```
Comment (pedrobaeza) [inline contract/models/res_partner.py]: Please use a common method called from both methods that uses an argument with the XML-ID, but don't repeat the very similar code.

File: contract/views/account_analytic_account_view.xml [hunk 10]
```diff
@@ -95,6 +98,57 @@
         </field>
     </record>
 
+    <record id="account_analytic_account_sale_form" model="ir.ui.view">
+        <field name="name">account.analytic.account.sale.form</field>
+        <field name="model">account.analytic.account</field>
+        <field name="inherit_id" ref="contract.account_analytic_account_recurring_form_form"/>
```
Comment (pedrobaeza) [inline contract/views/account_analytic_account_view.xml]: No need of putting `contract.`

File: contract/views/account_analytic_account_view.xml [hunk 11]
```diff
@@ -95,6 +98,57 @@
...
+        <field name="mode">primary</field>
+        <field name="priority" eval="20"/>
+        <field name="arch" type="xml">
+            <field name="partner_id" position="attributes">
+                <attribute name="string">Customer</attribute>
+                <attribute name="domain">[('customer', '=', True)]</attribute>
+                <attribute name="context">{
```
Comment (pedrobaeza) [inline contract/views/account_analytic_account_view.xml]: Don't put values in multi-line. They are read very ugly on UI.

File: contract/views/account_analytic_account_view.xml [hunk 12]
```diff
@@ -95,6 +98,57 @@
...
+            <field name="product_id" position="attributes">
+                <attribute name="domain">[('sale_ok', '=', True)]</attribute>
+            </field>
+        </field>
+    </record>
+
+    <record id="account_analytic_account_purchase_form" model="ir.ui.view">
+        <field name="name">account.analytic.account.purchase.form</field>
+        <field name="model">account.analytic.account</field>
+        <field name="inherit_id" ref="contract.account_analytic_account_recurring_form_form"/>
```
Comment (pedrobaeza) [inline contract/views/account_analytic_account_view.xml]: No need of putting `contract.`

File: contract/views/account_analytic_account_view.xml [hunk 13]
```diff
@@ -95,6 +98,57 @@
...
+        <field name="mode">primary</field>
+        <field name="priority" eval="20"/>
+        <field name="arch" type="xml">
+            <field name="partner_id" position="attributes">
+                <attribute name="string">Supplier</attribute>
+                <attribute name="domain">[('supplier', '=', True)]</attribute>
+                <attribute name="context">{
```
Comment (pedrobaeza) [inline contract/views/account_analytic_account_view.xml]: Don't put values in multi-line. They are read very ugly on UI.

File: contract/models/account_analytic_contract.py [hunk 14]
```diff
@@ -73,11 +79,22 @@ class AccountAnalyticContract(models.Model):
         default=lambda self: self.env.user.company_id,
     )
 
+    @api.onchange('contract_type')
+    def _onchange_contract_type(self):
```
Comment (pedrobaeza) [inline contract/models/account_analytic_contract.py:83]: Is this a real case? You can't modify the contract type through UI, so no onchange is going to be called.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: contract/__manifest__.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply contract/__manifest__.py]: I can revert it if you like. But `account_invoicing` is the module that makes visible the invoicing menu. And without that menu visible, the `contract` module doesn't make sense.

File: contract/models/account_analytic_contract.py (hunk 14 above)
Comment (MiquelRForgeFlow) [reply contract/models/account_analytic_contract.py:83]: In the contract template form view, you can change the contract_type.

File: contract/__manifest__.py (hunk 3 above)
Comment (pedrobaeza) [reply contract/__manifest__.py]: Well, the problem is to change dependencies on the middle of the lifecycle. You can put a comment for changing it on v12

File: contract/models/account_analytic_contract.py (hunk 14 above)
Comment (pedrobaeza) [reply contract/models/account_analytic_contract.py:83]: I think the field should be always invisible and not changeable by UI, as you have 2 separate menus already, and imagine who weird would be to have a sale contract (Maybe a better name should be "Customer Contract"?), generate a customer invoice, and then change to purchase the contract...

**Comment by pedrobaeza** [comment]
The more I think about it, the more I'm sure you should change the nomenclature from `sale` to `customer` and from `purchase` to `vendor` or `supplier`. Let's do it correctly from the beginning.

File: contract/__manifest__.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply contract/__manifest__.py]: Oh, right. But v12 already merged <code>account_invoicing</code> to <code>account</code> 😂

File: contract/models/account_analytic_contract.py (hunk 14 above)
Comment (MiquelRForgeFlow) [reply contract/models/account_analytic_contract.py:83]: There are three views:

- the sale contract, where contract_type is invisible.
- the purchase contract, where contract_type is invisible.
- the contract template, where contract_type is visible and changeable. You don't generate invoices from contract templates.

File: contract/__manifest__.py (hunk 3 above)
Comment (pedrobaeza) [reply contract/__manifest__.py]: Hehe, there's nothing to do :rofl:

File: contract/models/res_partner.py (hunk 8 above)
Comment (MiquelRForgeFlow) [reply contract/models/res_partner.py]: If I put outside the loop, what I do about <code>('partner_id', 'child_of', partner.ids),</code>?

Comment (pedrobaeza) [reply contract/models/res_partner.py]: Uhm, interesting. You can add `commercial_partner_id` field and group by it (or even better, search directly by it).

Comment (MiquelRForgeFlow) [reply contract/models/res_partner.py]: But it's not the same. That will give different result if the partner is different from its commercial_partner_id

Comment (pedrobaeza) [reply contract/models/res_partner.py]: Why? You shouldn't have "grandsons" (Odoo doesn't allow it naturally). Any way, the initial intention of that was to cover all the hierarchy.

Comment (MiquelRForgeFlow) [reply contract/models/res_partner.py]: That's not what I mean. I mean that now the search is counting the partner contracts and their child contracts. But not their parent contracts. Thus, if I use the commercial_partner_id, for child partners they will have in the count their parent also.

File: contract/migrations/11.0.4.0.0/post-migration.py (hunk 5 above)
Comment (MiquelRForgeFlow) [reply contract/migrations/11.0.4.0.0/post-migration.py]: But this way they are already filled by the migration to avoid any rare case.

File: contract/models/account_analytic_contract.py (hunk 14 above)
Comment (pedrobaeza) [reply contract/models/account_analytic_contract.py:83]: OK, I see. Thanks for clarifying me this.

File: contract/migrations/11.0.4.0.0/post-migration.py (hunk 5 above)
Comment (pedrobaeza) [reply contract/migrations/11.0.4.0.0/post-migration.py]: There's no rare case. This will simply be filled by ORM default. Which case can be?

File: contract/README.rst [hunk 15]
```diff
@@ -25,10 +25,10 @@ Contracts Management - Recurring
 
 |badge1| |badge2| |badge3| |badge4| |badge5| 
 
-This module brings back the contracts management with recurring invoicing
-features. Also you can print and send by email contract report.
+This module enables contracts management with recurring
+invoicing functions. Also you can print and send by email contract report.
 
-In upstream Odoo, this functionality was moved into the Enterprise edition.
+It works for sale contract and purchase contracts.
```
Comment (pedrobaeza) [inline contract/README.rst]: s/sale/customer
s/purchase/vendor

File: contract/models/account_analytic_account.py [hunk 16]
```diff
@@ -199,28 +199,33 @@ def _prepare_invoice_line(self, line, invoice_id):
         return invoice_line_vals
 
     @api.multi
-    def _prepare_invoice(self):
+    def _prepare_invoice(self, journal=None):
         self.ensure_one()
         if not self.partner_id:
             raise ValidationError(
                 _("You must first select a Customer for Contract %s!") %
```
Comment (pedrobaeza) [inline contract/models/account_analytic_account.py]: This text should change according the type

File: contract/models/account_analytic_contract.py [hunk 17]
```diff
@@ -26,6 +25,12 @@ class AccountAnalyticContract(models.Model):
         comodel_name="res.partner",
         string="Partner (always False)",
     )
+    contract_type = fields.Selection(
+        selection=[
+            ('sale', 'Sale'),
```
Comment (pedrobaeza) [inline contract/models/account_analytic_contract.py]: customer/vendor?

Comment (pedrobaeza) [reply contract/models/account_analytic_contract.py]: on label part only (for not wasting the rest)

File: contract/readme/DESCRIPTION.rst [hunk 18]
```diff
@@ -1,4 +1,4 @@
-This module brings back the contracts management with recurring invoicing
-features. Also you can print and send by email contract report.
+This module enables contracts management with recurring
+invoicing functions. Also you can print and send by email contract report.
 
-In upstream Odoo, this functionality was moved into the Enterprise edition.
+It works for sale contract and purchase contracts.
```
Comment (pedrobaeza) [inline contract/readme/DESCRIPTION.rst]: s/sale/customer
s/purchase/vendor

File: contract/views/account_analytic_account_view.xml [hunk 19]
```diff
@@ -95,6 +98,51 @@
...
+
+    <record id="account_analytic_account_purchase_form" model="ir.ui.view">
+        <field name="name">account.analytic.account.purchase.form</field>
+        <field name="model">account.analytic.account</field>
+        <field name="inherit_id" ref="account_analytic_account_recurring_form_form"/>
+        <field name="mode">primary</field>
+        <field name="priority" eval="20"/>
+        <field name="arch" type="xml">
+            <field name="partner_id" position="attributes">
+                <attribute name="string">Supplier</attribute>
```
Comment (pedrobaeza) [inline contract/views/account_analytic_account_view.xml:130]: Vendor for being aligned with new Accounting nomenclature?

File: contract/models/res_partner.py [hunk 20]
```diff
@@ -7,30 +7,44 @@
...
-                ('date_end', '=', False),
-                ('date_end', '>=', today),
-            ])
+            partner_child_ids = partner.child_ids.ids + partner.ids
+            partner.sale_contract_count = sum([
+                r[2] for r in result
+                if r[0] in partner_child_ids and r[1] == 'sale'])
+            partner.purchase_contract_count = sum([
+                r[2] for r in result
+                if r[0] in partner_child_ids and r[1] == 'purchase'])
```
Comment (pedrobaeza) [inline contract/models/res_partner.py:39]: Very good solution! :clap:

File: contract/views/account_analytic_account_view.xml (hunk 19 above)
Comment (MiquelRForgeFlow) [reply contract/views/account_analytic_account_view.xml:130]: But in all the module I used the Supplier nomenclature...

**Comment by MiquelRForgeFlow** [comment]
Any more comments @pedrobaeza ? 😎

**Comment by pedrobaeza** [comment]
It would be good to go for full coverage :wink:

File: contract/migrations/11.0.4.0.0/post-migration.py (hunk 5 above)
Comment (pedrobaeza) [reply contract/migrations/11.0.4.0.0/post-migration.py]: This is really not needed.

File: contract/models/res_partner.py [hunk 21]
```diff
@@ -7,30 +7,44 @@
...
+        string='Purchase Contracts',
         compute='_compute_contract_count',
     )
 
     def _compute_contract_count(self):
-        Contract = self.env['account.analytic.account']
+        contract_model = self.env['account.analytic.account']
         today = fields.Date.today()
+        fetch_data = contract_model.read_group([
+            ('recurring_invoices', '=', True),
```
Comment (pedrobaeza) [inline contract/models/res_partner.py:23]: I'm seeing that this doesn't filter by current partners, on the selection, so it's going to be low performance also. You need to get before all the partners (parent and children) for filtering the `read_group` statement.

Comment (MiquelRForgeFlow) [reply contract/models/res_partner.py:23]: done

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza codecov green :green_apple: 😎

**Comment by pedrobaeza** [comment]
@mreficent codecov is complaining about CI failed. Maybe a rebase? I insist that the migration script is not required.

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza No, it's just an error of the runbot. the same as https://github.com/OCA/runbot-addons/issues/173

**Comment by pedrobaeza** [comment]
OK, but please remove the migration script commit.

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza now codecov it's green and the migration commit removed

**Review by pedrobaeza — APPROVED** [review]
